Various tests + Poison Heal fix + Transistor update

Bassoonian opened the pull request:
Replaces #3181 

## Description
Adds battle tests for Defeatist, Gale Wings, Poison Heal, Sap Sipper, Earth Eater, Steelworker, Transistor, Dragon's Maw and Rocky Payload. Fixes Poison Heal ability popup and adds a config for Transistor's Gen9 nerf.

## **Discord contact info**
bassoonian

AsparagusEduardo commented on test/battle/ability/defeatist.c:
```diff
@@ -0,0 +1,46 @@
+    } FINALLY {
+        EXPECT_MUL_EQ(results[0].damage, Q_4_12(0.5), results[1].damage);
+    }
+}
```
New line at the end of the file

AsparagusEduardo commented on test/battle/ability/gale_wings.c:
```diff
@@ -0,0 +1,47 @@
+    } WHEN {
+        TURN { MOVE(player, MOVE_AERIAL_ACE); }
+    } SCENE {
+        if (hp == 100) {
+            MESSAGE("Talonflame used Aerial Ace!");
+        }
+        else {
+            MESSAGE("Foe Wobbuffet used Celebrate!");
+        }
```
This doesn't properly show that one message happens after the other.
A single `MESSAGE` just checks if the message happens at all.
```suggestion
        if (hp == 100) {
            MESSAGE("Talonflame used Aerial Ace!");
            MESSAGE("Foe Wobbuffet used Celebrate!");
        }
        else {
            MESSAGE("Talonflame used Aerial Ace!");
            MESSAGE("Foe Wobbuffet used Celebrate!");
        }
```

AsparagusEduardo commented on test/battle/ability/gale_wings.c:
```diff
@@ -0,0 +1,47 @@
+    } WHEN {
+        TURN { MOVE(player, move); }
+    } SCENE {
+        if (move == MOVE_AERIAL_ACE) {
+            MESSAGE("Talonflame used Aerial Ace!");
+        }
+        else {
+            MESSAGE("Foe Wobbuffet used Celebrate!");
+        }
```
Same as above.
```suggestion
        if (move == MOVE_AERIAL_ACE) {
            MESSAGE("Talonflame used Aerial Ace!");
            MESSAGE("Foe Wobbuffet used Celebrate!");
        }
        else {
            MESSAGE("Foe Wobbuffet used Celebrate!");
            MESSAGE("Talonflame used Aerial Ace!");
        }
```

AsparagusEduardo commented on test/battle/ability/gale_wings.c:
```diff
@@ -0,0 +1,47 @@
+            MESSAGE("Foe Wobbuffet used Celebrate!");
+        }
+    }
+}
```
```suggestion
}

TO_DO_BATTLE_TEST("Gale Wings doesn't increase priority of Flying-type Hidden Power, Natural Gift, Judgment or Tera Blast");
```

AsparagusEduardo commented on test/battle/ability/poison_heal.c:
```diff
@@ -0,0 +1,62 @@
+        NONE_OF {
+            ABILITY_POPUP(player, ABILITY_POISON_HEAL);
+            MESSAGE("The poisoning healed Shroomish a little bit!");
+            HP_BAR(player, damage: -50);
```
```suggestion
            HP_BAR(player, damage: -50);
            HP_BAR(player, damage: 50);
```

AsparagusEduardo commented on test/battle/ability/poison_heal.c:
```diff
@@ -0,0 +1,62 @@
+    }
+}
+
+SINGLE_BATTLE_TEST("Poison Heal heals from Toxic Poison damage")
```
This test should account for the fact that Toxic Poison heals at the same rate as regular Poison.

AsparagusEduardo commented on test/battle/ability/poison_heal.c:
```diff
@@ -0,0 +1,62 @@
+            HP_BAR(player, damage: -50);
+        }
+    }
+}
```
New line at the end of file.

AsparagusEduardo commented on test/battle/ability/sap_sipper.c:
```diff
@@ -0,0 +1,55 @@
+        NONE_OF { ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_STATS_CHANGE, player); }
+        NONE_OF { MESSAGE("Marill's Attack rose!"); }
+    }
+}
```
New line at the end of file.

Bassoonian commented:
Ready for re-review. Also added tests for Earth Eater as well as Steelworker and its variants, and added a config for Transistor (which is nerfed in Gen9+).

AsparagusEduardo commented:
> Ready for re-review. Also added tests for Earth Eater as well as Steelworker and its variants, and added a config for Transistor (which is nerfed in Gen9+).

Update the description and title with the new changes.

Bassoonian commented:
Done!

mrgriffin commented on src/battle_util.c:
```diff
@@ -8964,7 +8964,11 @@ static inline u32 CalcMoveBasePowerAfterModifiers(u32 move, u32 battlerAtk, u32
     case ABILITY_TRANSISTOR:
         if (moveType == TYPE_ELECTRIC)
+        #if B_TRANSISTOR >= GEN_9
+            modifier = uq4_12_multiply(modifier, UQ_4_12(5325 / 4096));
```
Damn, this isn't quite `UQ_4_12(1.3)`—`1.3 * 4096` is `5324.8` which truncates down to `5324`. I wonder if we could change it to round :thinking:

Anyway, I won't ask you to do that! Just something for me to look at for a future PR :)

Bassoonian replied:
I did try 1.3 while having the actual division in the code, but the test failed so having it here was the only way to make it pass :( it's unfortunate indeed, but we could always just take artistic license and simplify it to 1.3

AsparagusEduardo replied:
I want to see where this goes before accepting it.

AsparagusEduardo commented on test/battle/ability/poison_heal.c:
```diff
@@ -0,0 +1,77 @@
+        PLAYER(SPECIES_SHROOMISH) { Ability(ABILITY_POISON_HEAL); Status1(STATUS1_TOXIC_POISON);  HP(1), MaxHP(400); }
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_CELEBRATE); }
+        TURN { MOVE(player, MOVE_CELEBRATE); }
```
```suggestion
        TURN { }
        TURN { }
```

AsparagusEduardo commented on test/battle/ability/poison_heal.c:
```diff
@@ -0,0 +1,77 @@
+SINGLE_BATTLE_TEST("Poison Heal heals from (Toxic) Poison damage")
+{
+    u8 status;
+    PARAMETRIZE { status=STATUS1_POISON; }
+    PARAMETRIZE { status=STATUS1_TOXIC_POISON; }
```
```suggestion
    PARAMETRIZE { status = STATUS1_POISON; }
    PARAMETRIZE { status = STATUS1_TOXIC_POISON; }
```

AsparagusEduardo commented on test/battle/ability/gale_wings.c:
```diff
@@ -0,0 +1,53 @@
+    GIVEN {
+        ASSUME(B_GALE_WINGS >= GEN_7);
+        ASSUME(gBattleMoves[MOVE_AERIAL_ACE].type == TYPE_FLYING);
+        ASSUME(gBattleMoves[MOVE_FLARE_BLITZ].type == TYPE_FIRE);
```
```suggestion
        ASSUME(gBattleMoves[MOVE_FLARE_BLITZ].type != TYPE_FLYING);
```

AsparagusEduardo commented on test/battle/ability/dragons_maw.c:
```diff
@@ -0,0 +1,33 @@
+    PARAMETRIZE { move = MOVE_DRAGON_BREATH; ability = ABILITY_DRAGONS_MAW; }
+
+    GIVEN {
+        ASSUME(gBattleMoves[MOVE_TACKLE].type == TYPE_NORMAL);
```
```suggestion
        ASSUME(gBattleMoves[MOVE_TACKLE].type != TYPE_DRAGON);
```

Bassoonian commented:
Should be good to go other than the 5325/4096 multiplier, which I'll leave for @AsparagusEduardo to figure out :)

AsparagusEduardo commented:
*crap, the transistor update should've gone to `upcoming`*
